cmd/put.go

```diff
@@ -60,68 +62,96 @@ func uploadChunked(dbx files.Client, r io.Reader, commitInfo *files.CommitInfo,
+	destination, err := cmd.Flags().GetString("destination")
 	if err != nil {
-		return
+		return err
 	}
 
-	contentsInfo, err := contents.Stat()
-	if err != nil {
-		return
-	}
-
-	progressbar := &ioprogress.Reader{
-		Reader: contents,
-		DrawFunc: ioprogress.DrawTerminalf(os.Stderr, func(progress, total int64) string {
-			return fmt.Sprintf("Uploading %s/%s",
-				humanize.IBytes(uint64(progress)), humanize.IBytes(uint64(total)))
-		}),
-		Size: contentsInfo.Size(),
-	}
-
-	commitInfo := files.NewCommitInfo(dst)
-	commitInfo.Mode.Tag = "overwrite"
-
-	// The Dropbox API only accepts timestamps in UTC with second precision.
-	commitInfo.ClientModified = time.Now().UTC().Round(time.Second)
-
-	dbx := files.New(config)
-	if contentsInfo.Size() > chunkSize {
-		return uploadChunked(dbx, progressbar, commitInfo, contentsInfo.Size())
+	var waitGroup sync.WaitGroup
+	for _, arg := range args {
+		waitGroup.Add(1)
+		go func(arg string) error {
+			defer waitGroup.Done()
+			dst := "/" + path.Base(arg)
+
+			if destination != "" {
+				dst, err = validatePath(fullName(arg, destination))
+				if err != nil {
+					return err
+				}
+			}
+
+			contents, err := os.Open(arg)
+			defer contents.Close()
+			if err != nil {
+				return err
+			}
+
+			contentsInfo, err := contents.Stat()
+			if err != nil {
+				return err
+			}
+
+			progressbar := &ioprogress.Reader{
+				Reader: contents,
+				DrawFunc: ioprogress.DrawTerminalf(os.Stderr, func(progress, total int64) string {
+					return fmt.Sprintf("Uploading %s/%s",
+						humanize.IBytes(uint64(progress)), humanize.IBytes(uint64(total)))
+				}),
+				Size: contentsInfo.Size(),
+			}
+
+			commitInfo := files.NewCommitInfo(dst)
+			commitInfo.Mode.Tag = "overwrite"
+
+			// The Dropbox API only accepts timestamps in UTC with second precision.
+			commitInfo.ClientModified = time.Now().UTC().Round(time.Second)
+
+			dbx := files.New(config)
+			if contentsInfo.Size() > chunkSize {
+				err = uploadChunked(dbx, progressbar, commitInfo, contentsInfo.Size())
+				if err != nil {
+					return err
+				}
+			}
+
+			if uploadFile(dbx, commitInfo, progressbar) != nil {
+				return fmt.Errorf("Did not upload %s", arg)
+			}
+
+			return nil
+		}(arg)
 	}
+	waitGroup.Wait()
+	return nil
+}
 
-	if _, err = dbx.Upload(commitInfo, progressbar); err != nil {
-		return
+func uploadFile(dbx files.Client, commitInfo *files.CommitInfo, progressbar *ioprogress.Reader) error {
+	if _, err := dbx.Upload(commitInfo, progressbar); err != nil {
+		return err
 	}
+	return nil
+}
 
-	return
+func fullName(fileName, destination string) string {
+	return strings.TrimSuffix(destination, "/") + "/" + fileName
 }
 
 // putCmd represents the put command
 var putCmd = &cobra.Command{
-	Use:   "put [flags] <source> [<target>]",
+	Use:   "put [flags] <source>",
 	Short: "Upload files",
 	RunE:  put,
 }
 
 func init() {
 	RootCmd.AddCommand(putCmd)
+	putCmd.Flags().StringP("destination", "d", "", "specify a destination")
```

Comment:
My concern was that if you specify multiple arguments, how would we parse what the destination is? If we want to have the files uploaded to the root it wouldn't make sense to have the last argument be the destination. 

I'm not familiar with the S3 command line tools, what does the cli syntax look like?